### PR TITLE
Add GitHub Action to run Serverless integration tests on relevant PRs

### PR DESCRIPTION
**Pull request by nhinsch**
### What does this PR do?

We want to run the integration tests for the Datadog Lambda Extension (aka Serverless Agent) as a GitHub Action on pushes that affect the relevant files.

**This action won't be run on pushes that don't touch the serverless-related files.**

Before this PR works:

- [ ] We'll need to add the Serverless team's testing AWS credentials to the agent repository as secrets.
- [x] This PR needs to be merged: https://github.com/DataDog/datadog-lambda-extension/pull/42

The Lambda Extension uses a totally separate build and release process from the main agent, so this action is completely standalone and isn't written as a step in the GitLab CI pipeline.

### Motivation

By running the integration tests whenever we make changes, we can catch issues faster and more reliably. Exercising the test suite regularly will also force us to keep it up-to-date.

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] The appropriate `team/..` label has been applied, if known.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.

### CODE REVIEW

File: .github/workflows/serverless-integration.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
+name: "Serverless Integration Tests"
+
+on:
+  push:
+    paths:
+      - 'cmd/serverless/**'
+      - 'pkg/serverless/**'
+      - 'test/integration/serverless/**'
```
Comment (maxday) [inline .github/workflows/serverless-integration.yml:8]: I would add the current file `.github/workflows/serverless-integration.yml` as well? As if this changes, we might want to run integration tests as well?

File: .github/workflows/serverless-integration.yml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+      - name: Checkout the datadog-lambda-extension repository
+        uses: actions/checkout@v2
+        with:
+          repository: DataDog/datadog-lambda-extension
+          path: go/src/github.com/DataDog/datadog-lambda-extension
+
+      - name: Run tests
+        env:
+          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
```
Comment (maxday) [inline .github/workflows/serverless-integration.yml:36]: do we want to indicate that those secrets are serverless-related by adding a `SERVERLESS` prefix to those secrets?

Comment (nhinsch) [reply .github/workflows/serverless-integration.yml:36]: Excellent idea!

File: .github/workflows/serverless-integration.yml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
```
Comment (mx-psi) [inline .github/workflows/serverless-integration.yml:37]: Secrets will not be accessible to Dependabot or external contributors. We should skip the tests when secrets are not available, to avoid broken CI

File: .github/workflows/serverless-integration.yml [hunk 4]
```diff
@@ -0,0 +1,39 @@
+name: "Serverless Integration Tests"
```
Comment (mx-psi) [inline .github/workflows/serverless-integration.yml:1]: Can you add the serverless team as a codeowner of this file?

File: .github/workflows/serverless-integration.yml (hunk 1 above)
Comment (mx-psi) [inline .github/workflows/serverless-integration.yml:8]: What about a change in a package that the serverless build depends on (e.g. a change on the trace agent or the aggregator)? Can that break tests?

File: .github/workflows/serverless-integration.yml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout datadog-agent repository
+        uses: actions/checkout@v2
+        with:
+          path: go/src/github.com/DataDog/datadog-agent
```
Comment (mx-psi) [inline .github/workflows/serverless-integration.yml:18]:
```suggestion
          fetch-depth: 0
          path: go/src/github.com/DataDog/datadog-agent
```

**Review by mx-psi — COMMENTED** [review]
In general, we don't use Github Actions if we can avoid it: we use 3 other CI systems (Gitlab, CircleCI and AppVeyor) and try to add tests on these platforms if possible. Here, it makes sense to use GH if we want to run only when certain files change, but I am not convinced that the files currently listed on the PR are going to make for a useful CI in practice. 

I reviewed the PR assuming we stick to Github Actions but, if we don't think the CI rules are going to detect relevant PRs reliably then we want to run this on CircleCI or Gitlab.

File: .github/workflows/serverless-integration.yml (hunk 5 above)
Comment (nhinsch) [reply .github/workflows/serverless-integration.yml:18]: Per docs:

> Number of commits to fetch. 0 indicates all history for all branches and tags. Default: 1

I think we want to use the default value (1), no?

File: .github/workflows/serverless-integration.yml (hunk 1 above)
Comment (nhinsch) [reply .github/workflows/serverless-integration.yml:8]: Yes, it could, but it would have to be a pretty serious breaking API change to affect our tests.

**Comment by nhinsch** [comment]
Thanks for your feedback!

> Here, it makes sense to use GH if we want to run only when certain files change, but I am not convinced that the files currently listed on the PR are going to make for a useful CI in practice.

Our hope is that in practice, the vast majority of the changes that could break our tests would be touching files in one of these directories.

We don't feel these tests are ready to run on every single agent push at this time. Because these tests deploy resources to AWS, we might want to create a separate AWS account for this test suite to avoid hitting AWS API rate limits, which are already regular occurrences in the shared sandbox account. Also, because the tests are snapshot based, they need to be re-run locally and updated whenever there is a failure. Before we can make these tests part of the CI pipeline, we would probably need to rethink the test suite.

tl;dr We know this isn't true CI, but it will still be very helpful to our team in the interim. Running these tests automatically against our branches as we work on the Serverless agent is greatly preferable to running them manually, even if it's not going to immediately catch every issue in the agent that affects us.

File: .github/workflows/serverless-integration.yml (hunk 1 above)
Comment (nhinsch) [reply .github/workflows/serverless-integration.yml:8]: (see other comment for rationale)

File: .github/workflows/serverless-integration.yml (hunk 3 above)
Comment (nhinsch) [reply .github/workflows/serverless-integration.yml]: I was not able to skip the tests in the workflow definition, because the `secrets` context is only available in very limited places (like setting environment variables) and not in an `if` condition. Instead, I updated the test script to `exit 0` when credentials are not found.

Comment (mx-psi) [reply .github/workflows/serverless-integration.yml]: Works for me!

File: .github/workflows/serverless-integration.yml (hunk 5 above)
Comment (mx-psi) [reply .github/workflows/serverless-integration.yml:19]: Got confused by the CodeQL changes, when, in fact, we should fix it there, not here 😅 We can disregard this

**Review by mx-psi — APPROVED** [review]
Thanks for the comments! If the AWS rate limit is a cause for concern, let's go with this approach and reevaluate if needed

**Review by maxday — APPROVED** [review]
